src/index/ivf_raft/ivf_raft.cuh

```diff
@@ -45,103 +49,111 @@
+            if (invalid_id) {  // TODO(wphicks): assuming the branch is worth it here;
+                               // should analyze perf.
+                atomicAdd_block(invalid_count, 1);
+            }
+            invalid_id |= cur_id == std::numeric_limits<int64_t>::max();
+            ids[elem_index] = int(invalid_id) * -1 + int(!invalid_id) * ids[elem_index];
         }
-        offset *= 2;
-    }
-    if (tx == 0) {
-        temp[n - 1] = 0;
-    }
-    for (int d = 1; d < n; d *= 2) {
-        offset >>= 1;
         __syncthreads();
-        if (tx < d) {
-            int ai = offset * (2 * tx + 1) - 1;
-            int bi = offset * (2 * tx + 2) - 1;
-            int32_t t = temp[ai];
-            temp[ai] = temp[bi];
-            temp[bi] += t;
+
+        // Now move all valid results to the front of the row
+        auto cur_valid_index = row_index * k;
+        for (auto col_index = 0; col_index < k; ++col_index) {
+            auto elem_index = row_index * k + col_index;
+            // Just do one row per block for now; can improve this later
+            if (ids[elem_index] != -1 && threadIdx.x == 0) {
+                // Swap valid id to an earlier place in the row
+                ids[cur_valid_index] = ids[elem_index];
+                if (elem_index != cur_valid_index) {
+                    ids[elem_index] = -1;
+                    // Only count elements invalidated by the bitset. These are the
+                    // elements that will require a swap as opposed to just appearing at
+                    // the end of the row
+                }
+                dists[cur_valid_index++] = dists[elem_index];
+            }
+        }
+        // Check if we have enough valid results
+        if (threadIdx.x == 0) {
+            enough_valid[row_index] = (k - invalid_count[0] >= target_k);
         }
     }
-    __syncthreads();
-    g_odata[2 * tx + ty * n] += temp[2 * tx];
-    g_odata[2 * tx + 1 + ty * n] += temp[2 * tx + 1];
 }
 
+template <typename T, typename IndexT>
 __global__ void
-write_back(const int32_t* pre_ids_block, int64_t* ids_before, float* dis_before, int64_t* ids_after, float* dis_after,
-           int k1, int k2) {
-    int tx = blockIdx.x * blockDim.x + threadIdx.x;
-    int ty = blockIdx.y * blockDim.y + threadIdx.y;
-
-    extern __shared__ int32_t temp[];
-    for (int i = tx; i < k2; i += k1) {
-        temp[i] = pre_ids_block[ty * k2 + i];
-    }
-    __syncthreads();
-    int left = 0;
-    int right = k2;
-    while (left < right) {
-        int mid = (left + right) / 2;
-        if (temp[mid] == tx + 1) {
-            right = mid;
-        } else if (temp[mid] > tx + 1) {
-            right = mid;
-        } else if (temp[mid] < tx + 1) {
-            left = mid + 1;
+slice(T* out, T* in, IndexT out_rows, IndexT out_cols, IndexT in_rows, IndexT in_cols) {
+    for (auto i = blockIdx.x * blockDim.x + threadIdx.x; i < out_rows * out_cols; ++i) {
```

Comment:
Is this for-loop condition set correctly? Thread #0 will do `out_rows * out_cols` iteration and Thread #1 will do `out_rows * out_cols - 1` iteration?

Reply:
Whoops! Can't believe I screwed up a grid-strided loop, sigh...